### PR TITLE
Add env variables to provide proxy cert and key

### PR DESCRIPTION
**Pull request by vicapow**
I want to use a proxy that requires a cert when using pkg and I can't do this today without this functionality

### CODE REVIEW

**Review by robertsLando — CHANGES_REQUESTED** [review]
Needs to be documented somewhere

File: lib/utils.ts [hunk 1]
```diff
@@ -2,24 +2,51 @@ import fetch from 'node-fetch';
...
   log.showProgress(0);
 
-  const proxy =
-    process.env.HTTPS_PROXY ??
-    process.env.https_proxy ??
-    process.env.HTTP_PROXY ??
-    process.env.http_proxy;
+  const proxyHost = getProxyHost();
+
+  let proxy: HttpsProxyAgentOptions | void;
```
Comment (baparham) [inline lib/utils.ts:34]: `let proxy: HttpsProxyAgentOptions | undefined;` seems more accurate here

File: lib/utils.ts [hunk 2]
```diff
@@ -2,24 +2,51 @@ import fetch from 'node-fetch';
...
+      proxy = url.parse(proxyHost);
+      const certPath = getProxyCert();
+      if (certPath) {
+        proxy.cert = fs.readFileSync(certPath);
+      }
+      const keyPath = getProxyKey();
+      if (keyPath) {
+        proxy.key = fs.readFileSync(keyPath);
+      }
+    }
```
Comment (baparham) [inline lib/utils.ts:46]: this entire `if` block is indented one too many levels

File: lib/utils.ts [hunk 3]
```diff
@@ -2,24 +2,51 @@ import fetch from 'node-fetch';
...
-    process.env.HTTPS_PROXY ??
-    process.env.https_proxy ??
-    process.env.HTTP_PROXY ??
-    process.env.http_proxy;
+  const proxyHost = getProxyHost();
+
+  let proxy: HttpsProxyAgentOptions | void;
+  
+    if (proxyHost) {
+      proxy = url.parse(proxyHost);
```
Comment (baparham) [inline lib/utils.ts:37]: `url.parse` is deprecated, use `new URL(proxyHost)` instead

**Review by baparham — COMMENTED** [review]
Thanks for the contribution!
A few comments, and more generally, lines 31-46 can be broken out into at least one more function, like `const fetchInitOptions = getFetchInitOptions()` or something, and needs updates to the docs, as previously mentioned.

File: lib/utils.ts (hunk 3 above)
Comment (vicapow) [reply lib/utils.ts:37]: I can do that if you'd like. I was a bit worried about changing existing behavior. I see that htpss-proxy-agent still is using url.parse: https://github.com/TooTallNate/node-https-proxy-agent/blob/master/src/agent.ts#L34

**Comment by vicapow** [comment]
> Needs to be documented somewhere

Sure, I can do this. Although I saw the previous env `HTTPS_PROXY` wasn't documented either. Wasn't sure if that was intentional.

**Comment by vicapow** [comment]
👀 🙏

File: lib/utils.ts (hunk 3 above)
Comment (baparham) [reply lib/utils.ts]: hm, I would not expect there to be a conflict, and I'd rather use non-deprecated functions. `url.parse` returns a new URL object anyway, so it should be a drop in replacement, and more secure from things like host spoofing ([according to Node.js at least](https://nodejs.org/api/url.html#urlparseurlstring-parsequerystring-slashesdenotehost))

Comment (baparham) [reply lib/utils.ts]: @vicapow and @robertsLando I'd still like to see this resolved, unless @robertsLando has strong opinions otherwise

**Comment by baparham** [comment]
@vicapow and @robertsLando there is still an unresolved discussion here about using the deprecated `url.parse`

**Comment by robertsLando** [comment]
I would just use `new URL` instead of url.parse

**Comment by vicapow** [comment]
If you think it's safe enough to just use `new URL` instead of `url.parse` I'm happy to make the change. That said, there are cases were `new URL()` breaks, but `url.parse` doesn't. ie., when using an ip address, and likely others I'm not aware of:

```
> new URL('192.168.1.0:8080')
Uncaught TypeError [ERR_INVALID_URL]: Invalid URL
    at __node_internal_captureLargerStackTrace (node:internal/errors:484:5)
    at new NodeError (node:internal/errors:393:5)
    at URL.onParseError (node:internal/url:565:9)
    at new URL (node:internal/url:645:5) {
  input: '192.168.1.0:8080',
  code: 'ERR_INVALID_URL'
}
```

imo, the safest behavior is to maintain compatibility with [node-http-proxy-agent](https://github.com/TooTallNate/node-http-proxy-agent/blob/master/src/agent.ts#L38) since that's the behavior today and tackle the problem of changing (or risking changing) behavior in another PR. It should also make bisecting and reverting easier if there are issues with the switch.

**Comment by robertsLando** [comment]
@vicapow If you check the result of the url.parse it doens't look ok to me, It says `protocol: '192.168.1.0:'` `host: '8080'`, `hostname: '8080'`.

An url should be full so it must include protocol at start, in your example the protocol is missing so the url.parse output should be considered invalid

**Comment by robertsLando** [comment]
For anyone that is enterested in this project, giving that I don't receive any feedback from Vercel anymore, I created my own forks of both pkg and pkg-fetch:

[yao-pkg/pkg-fetch](https://github.com/yao-pkg/pkg-fetch)
[yao-pkg/pkg](https://github.com/yao-pkg/pkg)

If someone is interested in helping me create a PR or move your active PR to those repositories.

Thanks 🙏🏼

**Pull request closed without merging**